framework/i18n/Formatter.php

```diff
@@ -1741,4 +1778,220 @@ protected function createNumberFormatter($style, $decimals = null, $options = []
 
         return $formatter;
     }
+
+    /**
+     * Checks if string representations of given value and its normalized version are different.
+     * @param string|float|int $value
+     * @param float|int $normalizedValue
+     * @return bool
+     * @since 2.0.16
+     */
+    protected function isNormalizedValueMispresented($value, $normalizedValue)
+    {
+        if (empty($value)) {
+            $value = 0;
+        }
+
+        return (string) $normalizedValue !== (string) $value;
+    }
+
+    /**
+     * Fallback for formatting value as a decimal number.
+     *
+     * Property [[decimalSeparator]] will be used to represent the decimal point. The value is rounded automatically
+     * to the defined decimal digits.
+     *
+     * @param string|int|float $value the value to be formatted.
+     * @param int $decimals the number of digits after the decimal point. The default value is `2`.
+     * @return string the formatted result.
+     * @see decimalSeparator
+     * @see thousandSeparator
+     * @since 2.0.16
+     */
+    protected function asDecimalStringFallback($value, $decimals = null)
+    {
+        if (empty($value)) {
+            $value = 0;
+        }
+
+        $value = (string) $value;
+
+        if (strpos($value, '.') !== false) {
+            $integerPart = substr($value, 0, strrpos($value, '.'));
+            $fractionalPart = substr($value, strrpos($value, '.') + 1);
+        } else {
+            $integerPart = $value;
+            $fractionalPart = null;
+        }
+
+        $decimalOutput = '';
+
+        if ($decimals === null) {
+            $decimals = 2;
+        }
+
+        $roundIntegerPart = false;
+
+        if ($decimals > 0) {
+            $decimalSeparator = $this->decimalSeparator;
+            if ($this->decimalSeparator === null) {
+                $decimalSeparator = '.';
+            }
+
+            if ($fractionalPart === null) {
+                $fractionalPart = str_repeat('0', $decimals);
+            } elseif (strlen($fractionalPart) > $decimals) {
+                $cursor = $decimals;
+
+                if ((int) substr($fractionalPart, $cursor, 1) >= 5) {
+                    while (--$cursor >= 0) {
+                        $round = false;
+                        $newDigit = (int) substr($fractionalPart, $cursor, 1) + 1;
+
+                        if ($newDigit === 10) {
+                            $newDigit = 0;
+                            $round = true;
+                        }
+
+                        $fractionalPart = substr($fractionalPart, 0, $cursor) . $newDigit . substr($fractionalPart, $cursor + 1);
+
+                        if (!$round) {
+                            break;
+                        }
+                    }
+                    if ($round) {
+                        $roundIntegerPart = true;
```

Comment:
Additional variable and this condition could be avoided by defining `$round = false` or `$carry = 0` before `if ($fractionalPart === null) {`.

Reply:
Done.